language/compiler/ir_to_bytecode/syntax/src/ast.rs

```diff
@@ -143,13 +145,30 @@ pub enum StructDefinitionFields {
 #[derive(Debug, Eq, Hash, Ord, PartialEq, PartialOrd, Clone)]
 pub struct FunctionName(String);
 
+/// The kind of a type
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Kind {
+    All,
+    Resource,
+    Unrestricted,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct KindConstraint(pub String, pub Kind);
```

Comment:
nit type parameter
And I'd probably use named fields 

```
pub struct TypeParameter {
  pub name: String,
  pub constraint: Kind,
}
```

Reply:
Agree with Todd on this one.